va/http.go

```diff
@@ -0,0 +1,424 @@
+		deadline = time.Now().Add(100 * time.Second)
+	} else {
+		// Set the context deadline slightly shorter than the HTTP deadline, so we
+		// get a useful error rather than a generic "deadline exceeded" error. This
+		// lets us give a more specific error to the subscriber.
+		deadline = deadline.Add(-10 * time.Millisecond)
+	}
+	ctx, cancel := context.WithDeadline(ctx, deadline)
+	defer cancel()
+
+	// Create a target for the host, port and path
+	target, err := va.newHTTPValidationTarget(ctx, host, va.httpPort, path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Set up the initial validation request and a base validation record
+	initialReq, baseRecord, err := va.setupHTTPValidation(ctx, nil, target)
+	if err != nil {
+		return nil, []core.ValidationRecord{}, err
+	}
+
+	va.log.AuditInfof("Attempting to validate HTTP-01 for %q with GET to %q",
+		initialReq.Host, initialReq.URL.String())
+
+	// Create a closure around records & numRedirects we can use with a HTTP
+	// client to process redirects per our own policy (e.g. resolving IP
+	// addresses explicitly, not following redirects to ports != [80,443], etc)
+	records := []core.ValidationRecord{baseRecord}
+	numRedirects := 0
+	processRedirect := func(req *http.Request, via []*http.Request) error {
+		va.log.Infof("processing a HTTP redirect from the server to %q\n", req.URL.String())
+		// Only process up to maxRedirect redirects
+		if numRedirects > maxRedirect {
+			return berrors.ConnectionFailureError("Too many redirects")
+		}
+		numRedirects++
+		va.metrics.http01Redirects.Inc()
+
+		// Extract the redirect target's host and port. This will return an error if
+		// the redirect request scheme, host or port is not acceptable.
+		redirHost, redirPort, err := va.extractRequestTarget(req)
+		if err != nil {
+			return err
+		}
+
+		redirPath := req.URL.Path
+		// Setup a validation target for the redirect host. This will resolve IP
+		// addresses for the host explicitly.
+		redirTarget, err := va.newHTTPValidationTarget(ctx, redirHost, redirPort, redirPath)
+		if err != nil {
+			return err
+		}
+
+		// Mutate the existing redirect request to use a URL and Host based on the
+		// explicitly resolved target IPs. This will also give us a validationRecord
+		// for the redirect which we should append to the records.
+		_, redirRecord, err := va.setupHTTPValidation(ctx, req, redirTarget)
+		records = append(records, redirRecord)
+		if err != nil {
+			return err
+		}
+		va.log.Infof("following redirect to host %q url %q\n", req.Host, req.URL.String())
+		return nil
+	}
+
+	// Create a new HTTP client and check HTTP redirects it encounters with
+	// processRedirect
+	client := newHTTPClient()
+	client.CheckRedirect = processRedirect
+
+	// Make the initial validation request. This may result in redirects being
+	// followed.
+	httpResponse, err := client.Do(initialReq)
+	// If there was an error we may need to try a fallback.
+	if err != nil {
```

Comment:
This looks like any failure on the first request will trigger a fallback to IPv4. For instance, a failure to parse headers, or too many redirects. That's more permissive than our current code, and I'd like to maintain the current conservative semantics.

I think what we need to do to achieve that is to do a type assertion to `url.URLError` (the type `net/http` wraps its errors in), then do some additional type assertions on the inner error object (named `Err`) to see if it's a `net.Error`. Note: I'm not 100% certain of the details here. I'd recommend setting up a validation to a host that times out on IPv6 and doing some printf debugging of the exact error types.

Reply:
> This looks like any failure on the first request will trigger a fallback to IPv4. For instance, a failure to parse headers, or too many redirects. That's more permissive than our current code, and I'd like to maintain the current conservative semantics.

I thought we talked about this and reached the opposite conclusion. I thought it matched our current semantics because the dialer isn't checking the error type before falling back. Are you certain that neither of the conditions you mention would cause a fallback in the current code?

Reply:
> Are you certain that neither of the conditions you mention would cause a fallback in the current code?

I'm pretty certain. Remember that the fallback in the current code *only* exists within the dialer, so it can't know about HTTP-level errors.

I believe the conversation we had was along these lines: I recommended that we only fall back when the error is a timeout; you pointed out that the current custom dialer isn't that restrictive and will fall back on any dial; I agreed that it didn't make sense to make this logic more restrictive in this PR.

However, I do think it's worth not adding fallbacks for HTTP-level errors. I can see people accidentally winding up with a dependency on that behavior if, e.g. they have the wrong VirtualHost set up on their IPv6 interface.

Reply:
The majority of broken IPv6 I can remember from the forum fall into this bucket. I'm fine with matching the current behaviour but I don't think its super helpful if we aren't including HTTP-level errors. Maybe we should just not do any fallback and greatly simplify the code?

Reply:
> The majority of broken IPv6 I can remember from the forum fall into this bucket.

For "this bucket," you mean HTTP-level errors? That would make sense because the TCP-level errors are handled by our existing fallback.

> I'm fine with matching the current behaviour but I don't think its super helpful if we aren't including HTTP-level errors. Maybe we should just not do any fallback and greatly simplify the code?

I'm definitely open to removing our v6-to-v4 fallback, but I don't think we should do it as part of a refactor. I think that will make it harder to see if there are unrelated impacts. It also runs the risk of delaying the day we enable the feature flag, which in turn risks delaying the rollout of Go 1.11 builds.

Reply:
> For "this bucket," you mean HTTP-level errors? That would make sense because the TCP-level errors are handled by our existing fallback.

Yup, HTTP-level errors like an IPv6 webserver serving 404s for the challenge content. Good point that it would necessarily only be those kinds of errors we hear about because of the existing fallback code.

> I'm definitely open to removing our v6-to-v4 fallback, but I don't think we should do it as part of a refactor. I think that will make it harder to see if there are unrelated impacts. It also runs the risk of delaying the day we enable the feature flag, which in turn risks delaying the rollout of Go 1.11 builds.

These are all good points :+1: - I'll update the fallback to match the existing and we can revisit later if time permits.

Reply:
I think 006b618 is what we want here. I'm going to do some testing with a more realistic misconfigured IPv6 host to confirm.